Userland/Services/WindowServer/EventLoop.cpp

```diff
@@ -20,35 +22,94 @@ namespace WindowServer {
 
 EventLoop::EventLoop()
 {
-    m_keyboard_fd = open("/dev/input/keyboard/0", O_RDONLY | O_NONBLOCK | O_CLOEXEC);
-    m_mice_fd = open("/dev/input/mice", O_RDONLY | O_NONBLOCK | O_CLOEXEC);
-
     m_window_server = MUST(IPC::MultiServer<ConnectionFromClient>::try_create("/tmp/portal/window"));
     m_wm_server = MUST(IPC::MultiServer<WMConnectionFromClient>::try_create("/tmp/portal/wm"));
 
-    if (m_keyboard_fd >= 0) {
-        m_keyboard_notifier = Core::Notifier::construct(m_keyboard_fd, Core::Notifier::Type::Read);
-        m_keyboard_notifier->on_activation = [this] { drain_keyboard(); };
-    } else {
-        dbgln("Couldn't open /dev/input/keyboard/0");
+    m_mouse_devices_watcher = MUST(Core::FileWatcher::create());
+    m_keyboard_devices_watcher = MUST(Core::FileWatcher::create());
+
+    m_mouse_devices_watcher->on_change = [this](auto&) {
+        refresh_mouse_devices();
+    };
+
+    m_keyboard_devices_watcher->on_change = [this](auto&) {
+        refresh_keyboard_devices();
+    };
+
+    MUST(m_mouse_devices_watcher->add_watch("/tmp/system/devicemap/family/mouse/",
+        Core::FileWatcherEvent::Type::ChildCreated | Core::FileWatcherEvent::Type::ChildDeleted));
+
+    MUST(m_keyboard_devices_watcher->add_watch("/tmp/system/devicemap/family/keyboard/",
+        Core::FileWatcherEvent::Type::ChildCreated | Core::FileWatcherEvent::Type::ChildDeleted));
+
+    refresh_keyboard_devices();
+    refresh_mouse_devices();
+}
+
+void EventLoop::refresh_keyboard_devices()
+{
+    for (auto& device : m_keyboard_devices) {
+        close(device.fd);
+        device.notifier->close();
     }
+    m_keyboard_devices.clear();
+    Core::DirIterator di("/dev/input/keyboard/", Core::DirIterator::SkipParentAndBaseDir);
+    while (di.has_next()) {
+        auto path = di.next_path();
+        auto full_path = ByteString::formatted("/dev/input/keyboard/{}", path);
+        if (!FileSystem::is_device(full_path))
+            continue;
+
+        auto keyboard_fd_or_error = Core::System::open(full_path, O_RDONLY | O_NONBLOCK | O_CLOEXEC);
+        if (keyboard_fd_or_error.is_error()) {
+            dbgln("Couldn't open {}", full_path);
+            continue;
+        }
+
+        auto keyboard_fd = keyboard_fd_or_error.release_value();
+
+        auto keyboard_notifier = Core::Notifier::construct(keyboard_fd, Core::Notifier::Type::Read);
+        keyboard_notifier->on_activation = [this, keyboard_fd] { drain_keyboard(keyboard_fd); };
+        m_keyboard_devices.append({ keyboard_fd, keyboard_notifier });
+    }
+}
+
+void EventLoop::refresh_mouse_devices()
+{
```

Comment:
(nit: this looks very similar to the previous function. maybe both could call a common helper?)